src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int position;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+        position = -1;
+    }
+
     @Override
     public void add(T value) {
-
+        if (isNeedToResize()) {
+            grow();
+        }
+        elements[++position] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexToAdd(index);
+        if (isNeedToResize()) {
+            grow();
+        }
+        System.arraycopy(elements, index, elements, index + 1, size() - index);
+        elements[index] = value;
+        this.position++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexToGet(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexToGet(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexToGet(index);
+        T toRemove = (T) elements[index];
+        if (!isLastIndex(index)) {
+            System.arraycopy(elements, index + 1, elements, index, position - index);
+        }
+        elements[position--] = null;
+        return toRemove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        return remove(getIndexByValue(element));
     }
 
     @Override
     public int size() {
-        return 0;
+        return position + 1;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() <= 0;
+    }
+
+    private boolean isNeedToResize() {
+        return size() >= elements.length;
+    }
+
+    private void grow() {
+        int newLength = size() + (size() >> 1);
+        elements = Arrays.copyOf(elements, newLength);
```

Comment:
If i understand we only can use System.arraycopy, imagine we don't have Arrays class
```suggestion
        elements = System.arraycopy(...);
```

Reply:
ok, I changed to System.arraycopy , but there were no restrictions on the use Arrays.copyOf ))

Reply:
I understand, but it's a little weird if we implement ArrayList and use helper Arrays class. Perhaps I was confused by this paragraph in the common mistakes "Use System.arraycopy() to move your array elements"